lib/Conversion/TorchToLinalg/DataMovement.cpp

```diff
@@ -244,6 +244,283 @@ class ConvertAtenReflectionPad1dOp
+// operations.
+
+// To understand the lowering, consider this pytorch example:
+//
+// >>> t = torch.tensor([[[1.0,2,3],[4,5,6], [7,8,9]]])
+// >>> t
+// tensor([[[1., 2., 3.],
+//         [4., 5., 6.],
+//         [7., 8., 9.]]])
+// >>> torch.ops.aten.reflection_pad2d(t, [1,2,1,2])
+// tensor([[[5., 4., 5., 6., 5., 4.],
+//          [2., 1., 2., 3., 2., 1.],
+//          [5., 4., 5., 6., 5., 4.],
+//          [8., 7., 8., 9., 8., 7.],
+//          [5., 4., 5., 6., 5., 4.],
+//          [2., 1., 2., 3., 2., 1.]]])
+//
+// The result can be subdivided into "tiles" corresponding to either
+// the input tensor (in the center) or slices of the input tensor
+// whose width and height is determined by the padding sizes and which
+// are reflected through the side of the central input tensor that
+// they touch.
+// In the example above, the tiles are:
+// top left: [[5]]
+// top center: [[4,5,6]]
+// top right: [[5,4]]
+// center left [[2,1],[5,4],[8,7]]
+// center: copy of the input tensor
+// center right: [[2,1],[5,4],[8,7]]
+// bottom left: [[5,4],[2,1]]
+// center bottom: [[2,3,2]]
+// center right: [[2,1]]
+//
+// The lowering uses a tensor.extract_slice operation to create each tile,
+// a linalg.generic for the reflection, and a tensor.insert_slice to
+// insert the tile in the resulting tensor.
+class ConvertAtenReflectionPad2dOp
+    : public OpConversionPattern<AtenReflectionPad2dOp> {
+public:
+  using OpConversionPattern::OpConversionPattern;
+  LogicalResult
+  matchAndRewrite(AtenReflectionPad2dOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    if (failed(verifyLinalgCompatibleTypes(op, rewriter)))
+      return failure();
+
+    SmallVector<int64_t> padInts;
+    if (!matchPattern(op.getPadding(), m_TorchListOfConstantInts(padInts)))
+      return rewriter.notifyMatchFailure(
+          op, "only support constant int pad ranges");
+
+    Location loc = op.getLoc();
+    // Some generic helper functions for creating arithmetic operations.
+    auto createAdd = [&](Value x, Value y) {
+      return rewriter.create<arith::AddIOp>(loc, x, y);
+    };
+
+    auto createAdds = [&](std::initializer_list<Value> values) {
+      assert(values.size() >= 2);
+      return std::accumulate(values.begin() + 1, values.end(), data(values)[0],
+                             createAdd);
+    };
+
+    auto createSub = [&](Value x, Value y) {
+      return rewriter.create<arith::SubIOp>(loc, x, y);
+    };
+
+    auto createSubs = [&](std::initializer_list<Value> values) {
+      assert(values.size() >= 2);
+      return std::accumulate(values.begin() + 1, values.end(), data(values)[0],
+                             createSub);
+    };
+
+    // Enums for specifying the coordinates of a tile.  An "h" prefix
+    // is used to stand for "horizontal" and "v" for "vertical"
+    // throughout.
+    enum PadHLoc { LEFT = 0, RIGHT = 1, HCENTER = 2 };
```

Comment:
There are many places in the implementation where I am doing the same thing for the "vertical" and "horizontal" dimension. I first thought that it might improve readability to keep this distinction. I have now changed my mind and I believe that using two dimensional arrays to unify the code for the vertical and horizontal dimensions would simplify things. For instance, assignments like `extractHOffset[LEFT] = one;` and `extractVOffset[LEFT] = one;` below would be replaced by `extractOffset[dim][FIRST] = ONE` inside a loop that iterates the dimensions. But I have not made the change yet. Perhaps it is no necessary now? It would make sense to do this for the implementation of reflection_pad3d. I think the 2d and 1d operators could then be decomposed into it.

Reply:
I don't know the exact semantics of `reflection_pad3d` (and would have to see the decomposition you have in mind) but I would be surprised if a decomposition into pad3d would produce cleaner code than a direct lowering of 1d and 2d (fewer reshapes and such).

WRT the suggested changes here, I think it's fine to leave it as is and make the changes when adding support for 3d (if you're the one planning to implement support for 3d).